File: web/api/models/appointment.py
"""
appointment.py
Pydantic schemas — request/response models for the appointments API.
These are separate from the SQLAlchemy ORM models in db_models.py.
"""

from __future__ import annotations

from datetime import date, time
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class TherapyType(str, Enum):
    individual    = "individual"
    pareja        = "pareja"
    familia       = "familia"
    adolescentes  = "adolescentes"
    online        = "online"


class AppointmentStatus(str, Enum):
    pending   = "pending"
    confirmed = "confirmed"
    cancelled = "cancelled"


# ── Prices — single source of truth ──────────────────────────

PRICES: dict[TherapyType, int] = {
    TherapyType.individual:   600,
    TherapyType.pareja:       800,
    TherapyType.familia:      1000,
    TherapyType.adolescentes: 600,
    TherapyType.online:       600,
}


# ── Request schemas ───────────────────────────────────────────

class AppointmentCreate(BaseModel):
    patient_name:  str       = Field(..., min_length=2, max_length=120)
    patient_phone: str       = Field(..., min_length=7, max_length=20)
    patient_email: Optional[str] = Field(None, max_length=200)
    therapy_type:  TherapyType
    date:          date
    slot:          str       = Field(..., pattern=r"^\d{2}:\d{2}$",
                                     description="Hora de inicio HH:MM (24 h)")
    price:         int       = Field(..., gt=0)
    notes:         Optional[str] = Field(None, max_length=500)


class StatusUpdate(BaseModel):
    status: AppointmentStatus


# ── Response schemas ──────────────────────────────────────────

class AppointmentOut(BaseModel):
    id:              int
    patient_name:    str
    patient_phone:   str
    patient_email:   Optional[str]
    therapy_type:    TherapyType
    date:            date
    slot:            str
    status:          AppointmentStatus
    price:           int
    notes:           Optional[str]

    model_config = {"from_attributes": True}


# ── Availability schemas ──────────────────────────────────────

class DayStatus(str, Enum):
    available   = "available"
    limited     = "limited"
    unavailable = "unavailable"


class AvailableSlot(BaseModel):
    time:      str   # "HH:MM"
    available: bool


class SlotsResponse(BaseModel):
    date:  date
    slots: list[AvailableSlot]

File: mobile/lib/models/appointment.dart
class Appointment {
  final int    id;
  final String patientName;
  final String patientPhone;
  final String? patientEmail;
  final String therapyType;
  final String date;   // 'YYYY-MM-DD'
  final String slot;   // 'HH:MM'
  final String status; // 'pending' | 'confirmed' | 'cancelled'
  final int    price;
  final String? notes;

  const Appointment({
    required this.id,
    required this.patientName,
    required this.patientPhone,
    this.patientEmail,
    required this.therapyType,
    required this.date,
    required this.slot,
    required this.status,
    required this.price,
    this.notes,
  });

  factory Appointment.fromJson(Map<String, dynamic> j) => Appointment(
    id:           j['id'],
    patientName:  j['patient_name'],
    patientPhone: j['patient_phone'],
    patientEmail: j['patient_email'],
    therapyType:  j['therapy_type'],
    date:         j['date'],
    slot:         j['slot'],
    status:       j['status'],
    price:        j['price'],
    notes:        j['notes'],
  );

  Appointment copyWith({String? status}) => Appointment(
    id:           id,
    patientName:  patientName,
    patientPhone: patientPhone,
    patientEmail: patientEmail,
    therapyType:  therapyType,
    date:         date,
    slot:         slot,
    status:       status ?? this.status,
    price:        price,
    notes:        notes,
  );

  static const therapyLabels = {
    'individual':   'Terapia Individual',
    'pareja':       'Terapia en Pareja',
    'familia':      'Terapia Familiar',
    'adolescentes': 'Terapia para Adolescentes',
    'online':       'Terapia en Línea',
  };

  String get therapyLabel => therapyLabels[therapyType] ?? therapyType;

  String get formattedSlot {
    final h = int.parse(slot.split(':')[0]);
    if (h == 0)  return '12:00 AM';
    if (h < 12)  return '$h:00 AM';
    if (h == 12) return '12:00 PM';
    return '${h - 12}:00 PM';
  }
}
